File: e2e/graphql/devices_test.go
//go:build e2e

package graphql_test

import (
	"encoding/json"
	"testing"
	"time"

	"github.com/saffronjam/saffron-hive/e2e/infra"
)

func TestDevices_QueryAll(t *testing.T) {
	data, err := graphqlQuery(`{
		devices {
			id
			name
			type
			source
			available
		}
	}`, nil)
	if err != nil {
		t.Fatalf("query devices: %v", err)
	}

	var result struct {
		Devices []struct {
			ID        string `json:"id"`
			Name      string `json:"name"`
			Type      string `json:"type"`
			Source    string `json:"source"`
			Available bool   `json:"available"`
		} `json:"devices"`
	}
	if err := json.Unmarshal(data, &result); err != nil {
		t.Fatalf("unmarshal: %v", err)
	}

	if len(result.Devices) != expectedDeviceCount {
		t.Fatalf("expected %d devices, got %d", expectedDeviceCount, len(result.Devices))
	}

	nameSet := make(map[string]struct{}, len(result.Devices))
	for _, d := range result.Devices {
		nameSet[d.Name] = struct{}{}
		if d.Source != "zigbee" {
			t.Errorf("device %q source=%q, want zigbee", d.Name, d.Source)
		}
		if d.ID == "" {
			t.Errorf("device %q has empty ID", d.Name)
		}
	}

	for _, expected := range []string{"Living Room Light", "Bedroom Light", "Kitchen Light", "Living Room Sensor", "Outdoor Sensor", "Office Button", "Lava Lamp"} {
		if _, ok := nameSet[expected]; !ok {
			t.Errorf("expected device %q not found", expected)
		}
	}
}

func TestDevices_VerifyTypes(t *testing.T) {
	data, err := graphqlQuery(`{ devices { name type } }`, nil)
	if err != nil {
		t.Fatalf("query: %v", err)
	}

	var result struct {
		Devices []struct {
			Name string `json:"name"`
			Type string `json:"type"`
		} `json:"devices"`
	}
	if err := json.Unmarshal(data, &result); err != nil {
		t.Fatalf("unmarshal: %v", err)
	}

	expected := map[string]string{
		"Living Room Light":  "light",
		"Bedroom Light":      "light",
		"Kitchen Light":      "light",
		"Living Room Sensor": "sensor",
		"Outdoor Sensor":     "sensor",
		"Office Button":      "button",
		"Lava Lamp":          "plug",
	}

	for _, d := range result.Devices {
		if want, ok := expected[d.Name]; ok {
			if d.Type != want {
				t.Errorf("device %q type=%q, want %q", d.Name, d.Type, want)
			}
		}
	}
}

func TestDevices_StateChange(t *testing.T) {
	lightState, err := infra.LoadLightState()
	if err != nil {
		t.Fatalf("load fixture: %v", err)
	}

	if err := publisher.PublishDeviceState("Living Room Light", lightState); err != nil {
		t.Fatalf("publish state: %v", err)
	}

	ok := pollUntil(5*time.Second, 100*time.Millisecond, func() bool {
		data, err := graphqlQuery(`{
			devices {
				name
				state { on brightness colorTemp }
			}
		}`, nil)
		if err != nil {
			return false
		}
		var result struct {
			Devices []struct {
				Name  string `json:"name"`
				State struct {
					On         *bool `json:"on"`
					Brightness *int  `json:"brightness"`
					ColorTemp  *int  `json:"colorTemp"`
				} `json:"state"`
			} `json:"devices"`
		}
		if err := json.Unmarshal(data, &result); err != nil {
			return false
		}
		for _, d := range result.Devices {
			if d.Name == "Living Room Light" {
				return d.State.Brightness != nil && *d.State.Brightness == 200
			}
		}
		return false
	})

	if !ok {
		t.Fatal("timed out waiting for light state to propagate")
	}
}

func TestDevices_QuerySingleByID(t *testing.T) {
	deviceID, err := queryDeviceIDByName("Living Room Light")
	if err != nil {
		t.Fatalf("find device: %v", err)
	}

	data, err := graphqlQuery(`query($id: ID!) {
		device(id: $id) {
			id
			name
			type
			source
			available
		}
	}`, map[string]any{"id": deviceID})
	if err != nil {
		t.Fatalf("query device: %v", err)
	}

	var result struct {
		Device struct {
			ID        string `json:"id"`
			Name      string `json:"name"`
			Type      string `json:"type"`
			Source    string `json:"source"`
			Available bool   `json:"available"`
		} `json:"device"`
	}
	if err := json.Unmarshal(data, &result); err != nil {
		t.Fatalf("unmarshal: %v", err)
	}

	if result.Device.ID != deviceID {
		t.Errorf("id=%q, want %q", result.Device.ID, deviceID)
	}
	if result.Device.Name != "Living Room Light" {
		t.Errorf("name=%q, want %q", result.Device.Name, "Living Room Light")
	}
	if result.Device.Type != "light" {
		t.Errorf("type=%q, want light", result.Device.Type)
	}
	if result.Device.Source != "zigbee" {
		t.Errorf("source=%q, want zigbee", result.Device.Source)
	}
}

func TestDevices_QuerySingleByID_NotFound(t *testing.T) {
	data, err := graphqlQuery(`query($id: ID!) {
		device(id: $id) { id name }
	}`, map[string]any{"id": "nonexistent-device-id"})
	if err != nil {
		t.Fatalf("query: %v", err)
	}

	var result struct {
		Device *struct {
			ID   string `json:"id"`
			Name string `json:"name"`
		} `json:"device"`
	}
	if err := json.Unmarshal(data, &result); err != nil {
		t.Fatalf("unmarshal: %v", err)
	}
	if result.Device != nil {
		t.Errorf("expected null device, got %+v", result.Device)
	}
}

func TestDevices_SetDeviceState(t *testing.T) {
	deviceID, err := queryDeviceIDByName("Kitchen Light")
	if err != nil {
		t.Fatalf("find device: %v", err)
	}

	cmdCh, err := publisher.SubscribeCommands()
	if err != nil {
		t.Fatalf("subscribe commands: %v", err)
	}

	_, err = graphqlMutation(`mutation($deviceId: ID!, $state: DeviceStateInput!) {
		setDeviceState(deviceId: $deviceId, state: $state) { id name }
	}`, map[string]any{
		"deviceId": deviceID,
		"state": map[string]any{
			"on":         true,
			"brightness": 128,
		},
	})
	if err != nil {
		t.Fatalf("setDeviceState: %v", err)
	}

	ok := pollUntil(5*time.Second, 50*time.Millisecond, func() bool {
		select {
		case msg := <-cmdCh:
			if msg.Topic == "zigbee2mqtt/Kitchen Light/set" {
				return true
			}
		default:
		}
		return false
	})
	if !ok {
		t.Fatal("timed out waiting for MQTT command from setDeviceState")
	}
}

func TestDevices_UpdateDeviceName(t *testing.T) {
	// KNOWN BUG: updateDevice updates DB but mapDeviceFromReader reads name from
	// in-memory StateReader which still has the old name. Same class as group-target bugs.
	t.Skip("KNOWN BUG: updateDevice response reads from memory store, not DB")
	deviceID, err := queryDeviceIDByName("Bedroom Light")
	if err != nil {
		t.Fatalf("find device: %v", err)
	}

	data, err := graphqlMutation(`mutation($id: ID!, $input: UpdateDeviceInput!) {
		updateDevice(id: $id, input: $input) { id name }
	}`, map[string]any{
		"id":    deviceID,
		"input": map[string]any{"name": "Renamed Bedroom Light"},
	})
	if err != nil {
		t.Fatalf("updateDevice: %v", err)
	}

	var updateResult struct {
		UpdateDevice struct {
			ID   string `json:"id"`
			Name string `json:"name"`
		} `json:"updateDevice"`
	}
	if err := json.Unmarshal(data, &updateResult); err != nil {
		t.Fatalf("unmarshal: %v", err)
	}
	if updateResult.UpdateDevice.Name != "Renamed Bedroom Light" {
		t.Errorf("name=%q, want %q", updateResult.UpdateDevice.Name, "Renamed Bedroom Light")
	}

	queryData, err := graphqlQuery(`query($id: ID!) {
		device(id: $id) { id name }
	}`, map[string]any{"id": deviceID})
	if err != nil {
		t.Fatalf("query device: %v", err)
	}

	var queryResult struct {
		Device struct {
			Name string `json:"name"`
		} `json:"device"`
	}
	if err := json.Unmarshal(queryData, &queryResult); err != nil {
		t.Fatalf("unmarshal: %v", err)
	}
	if queryResult.Device.Name != "Renamed Bedroom Light" {
		t.Errorf("persisted name=%q, want %q", queryResult.Device.Name, "Renamed Bedroom Light")
	}

	_, err = graphqlMutation(`mutation($id: ID!, $input: UpdateDeviceInput!) {
		updateDevice(id: $id, input: $input) { id name }
	}`, map[string]any{
		"id":    deviceID,
		"input": map[string]any{"name": "Bedroom Light"},
	})
	if err != nil {
		t.Fatalf("restore name: %v", err)
	}
}

func TestDevices_SetDeviceState_InvalidID(t *testing.T) {
	err := graphqlMutationExpectError(`mutation($deviceId: ID!, $state: DeviceStateInput!) {
		setDeviceState(deviceId: $deviceId, state: $state) { id }
	}`, map[string]any{
		"deviceId": "nonexistent-device-id",
		"state": map[string]any{
			"on": true,
		},
	})
	if err != nil {
		t.Fatalf("expected GraphQL error for invalid device ID, got: %v", err)
	}
}

func TestDevices_AvailabilityChange(t *testing.T) {
	if err := publisher.PublishAvailability("Living Room Light", false); err != nil {
		t.Fatalf("publish availability: %v", err)
	}

	ok := pollUntil(5*time.Second, 100*time.Millisecond, func() bool {
		data, err := graphqlQuery(`{ devices { name available } }`, nil)
		if err != nil {
			return false
		}
		var result struct {
			Devices []struct {
				Name      string `json:"name"`
				Available bool   `json:"available"`
			} `json:"devices"`
		}
		if err := json.Unmarshal(data, &result); err != nil {
			return false
		}
		for _, d := range result.Devices {
			if d.Name == "Living Room Light" {
				return !d.Available
			}
		}
		return false
	})

	if !ok {
		t.Fatal("timed out waiting for availability change")
	}

	if err := publisher.PublishAvailability("Living Room Light", true); err != nil {
		t.Fatalf("restore availability: %v", err)
	}

	pollUntil(5*time.Second, 100*time.Millisecond, func() bool {
		data, err := graphqlQuery(`{ devices { name available } }`, nil)
		if err != nil {
			return false
		}
		var result struct {
			Devices []struct {
				Name      string `json:"name"`
				Available bool   `json:"available"`
			} `json:"devices"`
		}
		if err := json.Unmarshal(data, &result); err != nil {
			return false
		}
		for _, d := range result.Devices {
			if d.Name == "Living Room Light" {
				return d.Available
			}
		}
		return false
	})
}

File: web/e2e/devices.test.ts
import { describe, it, expect } from "vitest";
import { graphql } from "$lib/gql";
import { pipe, subscribe } from "wonka";
import {
  getContext,
  getBridgeDevicesFixture,
  publishDeviceState,
  getLightStateFixture,
  subscribeMQTTCommands,
} from "./setup.js";

const DEVICES_QUERY = graphql(`
  query E2EDevicesList {
    devices {
      id
      name
      friendlyName
      source
      type
      available
      state {
        on
        brightness
        colorTemp
        temperature
        humidity
        battery
        power
        voltage
        current
        energy
      }
    }
  }
`);

const DEVICE_QUERY = graphql(`
  query E2EDevice($id: ID!) {
    device(id: $id) {
      id
      name
      friendlyName
      source
      type
      available
      state {
        on
        brightness
        colorTemp
        temperature
        humidity
        battery
        power
        voltage
        current
        energy
      }
    }
  }
`);

const ZIGBEE_METADATA_QUERY = graphql(`
  query E2EZigbeeDeviceMetadata($id: ID!) {
    device(id: $id) {
      zigbee2Mqtt {
        imageCandidate
        imageVersion
        ieeeAddress
        networkAddress
        supported
        softwareBuildId
        definitionUrl
        definition {
          model
          vendor
          description
          supportsOta
        }
        ota {
          state
          installedVersion
          latestVersion
          progress
        }
        endpoints {
          id
          profileId
          deviceId
          inputClusters
          outputClusters
          bindings {
            cluster
            targetType
            targetIeeeAddress
            targetEndpoint
            targetGroupId
          }
          reportings {
            cluster
            attribute
            minimumReportInterval
            maximumReportInterval
            reportableChange
          }
        }
        groups {
          id
          providerGroupId
          name
          endpoint
        }
        bridgeInfo {
          adapterType
          firmwareVersion
          channel
          panId
          extendedPanId
          zigbee2MqttVersion
          zigbee2MqttCommit
          zigbeeHerdsmanVersion
          zigbeeHerdsmanConvertersVersion
        }
      }
    }
  }
`);

interface BridgeDevice {
  ieee_address: string;
  friendly_name: string;
  type: string;
}

const SET_DEVICE_STATE = graphql(`
  mutation E2ESetDeviceState($deviceId: ID!, $state: DeviceStateInput!) {
    setTargetState(targetType: DEVICE, targetId: $deviceId, state: $state)
  }
`);

const UPDATE_DEVICE = graphql(`
  mutation E2EUpdateDevice($id: ID!, $input: UpdateDeviceInput!) {
    updateDevice(id: $id, input: $input) {
      id
      name
    }
  }
`);

const DEVICE_STATE_CHANGED_SUB = graphql(`
  subscription E2EDevicesDeviceStateChanged {
    deviceStateChanged {
      deviceId
      state {
        on
        brightness
        colorTemp
      }
    }
  }
`);

const COORDINATOR_TYPE = "Coordinator";

describe("devices", () => {
  it("should return every device in the bridge registry", async () => {
    const { graphqlClient } = getContext();
    const fixtures = getBridgeDevicesFixture() as BridgeDevice[];

    const result = await graphqlClient.query(DEVICES_QUERY, {}).toPromise();

    expect(result.error).toBeUndefined();
    expect(result.data).toBeDefined();
    expect(result.data!.devices).toHaveLength(fixtures.length);
  });

  it("should register the coordinator as a hub device", async () => {
    const { graphqlClient } = getContext();
    const fixtures = getBridgeDevicesFixture() as BridgeDevice[];
    const coordinator = fixtures.find((d) => d.type === COORDINATOR_TYPE);

    const result = await graphqlClient.query(DEVICES_QUERY, {}).toPromise();

    const hub = result.data!.devices.find((d) => d.id === coordinator!.ieee_address);
    expect(hub).toBeDefined();
    expect(hub!.type).toBe("hub");
    expect(hub!.available).toBe(true);
  });

  it("should have correct device fields matching fixtures", async () => {
    const { graphqlClient } = getContext();
    const fixtures = getBridgeDevicesFixture() as BridgeDevice[];

    const result = await graphqlClient.query(DEVICES_QUERY, {}).toPromise();

    expect(result.data).toBeDefined();

    const deviceNames = result.data!.devices.map((d) => d.friendlyName);
    for (const fixture of fixtures) {
      expect(deviceNames).toContain(fixture.friendly_name);
    }

    for (const device of result.data!.devices) {
      expect(device.id).toBeTruthy();
      expect(device.friendlyName).toBeTruthy();
      expect(device.source).toBe("zigbee2mqtt");
      expect(typeof device.available).toBe("boolean");
    }
  });

  it("should query a single device by ID", async () => {
    const { graphqlClient } = getContext();

    const listResult = await graphqlClient.query(DEVICES_QUERY, {}).toPromise();

    expect(listResult.data).toBeDefined();
    const firstDevice = listResult.data!.devices[0];

    const result = await graphqlClient.query(DEVICE_QUERY, { id: firstDevice.id }).toPromise();

    expect(result.error).toBeUndefined();
    expect(result.data).toBeDefined();
    expect(result.data!.device).toBeDefined();
    expect(result.data!.device!.id).toBe(firstDevice.id);
    expect(result.data!.device!.friendlyName).toBe(firstDevice.friendlyName);
  });

  it("should return sanitized, typed Zigbee metadata only on device detail", async () => {
    const { graphqlClient } = getContext();
    await expect
      .poll(
        async () => {
          const result = await graphqlClient
            .query(
              ZIGBEE_METADATA_QUERY,
              { id: "0x54ef44100166fcae" },
              { requestPolicy: "network-only" },
            )
            .toPromise();
          return result.data?.device?.zigbee2Mqtt?.ieeeAddress ?? null;
        },
        { timeout: 10_000 },
      )
      .toBe("0x54ef44100166fcae");

    const t1 = await graphqlClient
      .query(ZIGBEE_METADATA_QUERY, { id: "0x54ef44100166fcae" }, { requestPolicy: "network-only" })
      .toPromise();
    expect(t1.error).toBeUndefined();
    expect(t1.data?.device?.zigbee2Mqtt).toMatchObject({
      ieeeAddress: "0x54ef44100166fcae",
      networkAddress: 2710,
      supported: true,
      softwareBuildId: "2019www.",
      definition: {
        model: "MCCGQ12LM",
        vendor: "Aqara",
        description: "Door and window sensor T1",
        supportsOta: false,
      },
    });
    expect(t1.data?.device?.zigbee2Mqtt?.endpoints).toHaveLength(1);
    expect(t1.data?.device?.zigbee2Mqtt?.imageCandidate).toBe(true);
    expect(t1.data?.device?.zigbee2Mqtt?.imageVersion).toMatch(/^[a-f0-9]{64}$/);
    expect(t1.data?.device?.zigbee2Mqtt?.definitionUrl).toBe(
      "https://www.zigbee2mqtt.io/devices/MCCGQ12LM.html",
    );

    const p100 = await graphqlClient
      .query(ZIGBEE_METADATA_QUERY, { id: "0x54ef4410015e4b68" }, { requestPolicy: "network-only" })
      .toPromise();
    expect(p100.error).toBeUndefined();
    expect(p100.data?.device?.zigbee2Mqtt?.softwareBuildId).toBeNull();
    expect(p100.data?.device?.zigbee2Mqtt?.endpoints).toHaveLength(2);
    expect(p100.data?.device?.zigbee2Mqtt?.endpoints[0].bindings).toHaveLength(2);
    expect(p100.data?.device?.zigbee2Mqtt?.endpoints[0].reportings).toHaveLength(1);

    const unsupported = await graphqlClient
      .query(ZIGBEE_METADATA_QUERY, { id: "0x00124b0000000001" }, { requestPolicy: "network-only" })
      .toPromise();
    expect(unsupported.error).toBeUndefined();
    expect(unsupported.data?.device?.zigbee2Mqtt?.supported).toBe(false);
    expect(unsupported.data?.device?.zigbee2Mqtt?.definition).toBeNull();
    expect(unsupported.data?.device?.zigbee2Mqtt?.endpoints).toEqual([]);

    const coordinator = await graphqlClient
      .query(ZIGBEE_METADATA_QUERY, { id: "0x00124b0000000000" }, { requestPolicy: "network-only" })
      .toPromise();
    expect(coordinator.error).toBeUndefined();
    expect(coordinator.data?.device?.zigbee2Mqtt?.bridgeInfo).toEqual({
      adapterType: "ZStack3x0",
      firmwareVersion: "20240710",
      channel: 20,
      panId: 6754,
      extendedPanId: "0x00124b000000abcd",
      zigbee2MqttVersion: "2.7.2",
      zigbee2MqttCommit: "a1b2c3d",
      zigbeeHerdsmanVersion: "6.1.4",
      zigbeeHerdsmanConvertersVersion: "25.30.0",
    });
  });

  it("should reflect state changes after MQTT publish", async () => {
    const { graphqlClient } = getContext();
    const lightState = getLightStateFixture();

    await publishDeviceState("Living Room Light", lightState);
    await new Promise((r) => setTimeout(r, 1000));

    const result = await graphqlClient.query(DEVICES_QUERY, {}).toPromise();

    expect(result.data).toBeDefined();
    const light = result.data!.devices.find((d) => d.friendlyName === "Living Room Light");
    expect(light).toBeDefined();
    expect(light!.state).toBeDefined();
  });

  it("should deliver state changes via subscription", async () => {
    const { graphqlClient } = getContext();

    const received = new Promise<{ deviceId: string }>((resolve, reject) => {
      const timeout = setTimeout(() => reject(new Error("Subscription timed out")), 10_000);

      const { unsubscribe } = pipe(
        graphqlClient.subscription(DEVICE_STATE_CHANGED_SUB, {}),
        subscribe((result) => {
          if (result.data) {
            clearTimeout(timeout);
            unsubscribe();
            resolve(result.data.deviceStateChanged);
          }
        }),
      );
    });

    await new Promise((r) => setTimeout(r, 500));
    await publishDeviceState("Living Room Light", {
      state: "ON",
      brightness: 150,
      color_temp: 300,
    });

    const event = await received;
    expect(event.deviceId).toBeTruthy();
  });

  it("should return null for nonexistent device ID", async () => {
    const { graphqlClient } = getContext();

    const result = await graphqlClient.query(DEVICE_QUERY, { id: "nonexistent" }).toPromise();

    expect(result.error).toBeUndefined();
    expect(result.data).toBeDefined();
    expect(result.data!.device).toBeNull();
  });

  it("should set device state via mutation", async () => {
    const { graphqlClient } = getContext();

    const listResult = await graphqlClient.query(DEVICES_QUERY, {}).toPromise();

    expect(listResult.data).toBeDefined();
    const lightDevice = listResult.data!.devices.find(
      (d) => d.friendlyName === "Living Room Light",
    );
    expect(lightDevice).toBeDefined();

    const { messages, cleanup } = await subscribeMQTTCommands();

    const result = await graphqlClient
      .mutation(SET_DEVICE_STATE, {
        deviceId: lightDevice!.id,
        state: { on: true, brightness: 200 },
      })
      .toPromise();

    expect(result.error).toBeUndefined();
    expect(result.data).toBeDefined();
    expect(result.data!.setTargetState).toBe(true);

    await new Promise((r) => setTimeout(r, 500));
    expect(messages.length).toBeGreaterThan(0);

    await cleanup();
  });

  it("should return error for setTargetState with invalid ID", async () => {
    const { graphqlClient } = getContext();

    const result = await graphqlClient
      .mutation(SET_DEVICE_STATE, {
        deviceId: "nonexistent",
        state: { on: true },
      })
      .toPromise();

    expect(result.error).toBeDefined();
  });

  // EXPECTED FAIL: updateDevice updates DB but response reads from in-memory StateReader
  // which still has the old name. Same class of bug as the group-target resolution issues.
  it.skip("should rename a device via updateDevice", async () => {
    const { graphqlClient } = getContext();

    const devicesResult = await graphqlClient.query(DEVICES_QUERY, {}).toPromise();
    expect(devicesResult.data).toBeDefined();
    expect(devicesResult.data!.devices.length).toBeGreaterThan(0);

    const device = devicesResult.data!.devices[0];
    const originalName = device.name;
    const newName = `Renamed ${Date.now()}`;

    const updateResult = await graphqlClient
      .mutation(UPDATE_DEVICE, {
        id: device.id,
        input: { name: newName },
      })
      .toPromise();

    expect(updateResult.error).toBeUndefined();
    expect(updateResult.data).toBeDefined();
    expect(updateResult.data!.updateDevice.name).toBe(newName);

    const queryResult = await graphqlClient.query(DEVICE_QUERY, { id: device.id }).toPromise();

    expect(queryResult.data).toBeDefined();
    expect(queryResult.data!.device).toBeDefined();
    expect(queryResult.data!.device!.name).toBe(newName);

    await graphqlClient
      .mutation(UPDATE_DEVICE, {
        id: device.id,
        input: { name: originalName },
      })
      .toPromise();
  });
});
